closed/src/java.base/share/native/libjli/criuhelpers.c

```diff
@@ -0,0 +1,334 @@
+        char logLevelString[2] = { 0 };
+        sprintf(logLevelString, "%d", logLevel);
+        JLI_ReportErrorMessage("Failed to allocate memory for the command option, value=%s, format=%s.", logLevelString, "-v%d");
+        restoreStatus = -1;
+        goto done;
+    }
+    if (snprintf(logLevelOption, length + 1, "-v%d", logLevel) < 0) {
+        char logLevelString[2] = { 0 };
+        sprintf(logLevelString, "%d", logLevel);
+        JLI_ReportErrorMessage("Failed to allocate memory for the command option, value=%s, format=%s.", logLevelString, "-v%d");
+        restoreStatus = -1;
+        goto freeLogLevelOption;
+    }
+    argv[argc++] = logLevelOption;
+    argv[argc++] = "--shell-job";
+    if (unprivilegedModeOn) {
+        argv[argc++] = "--unprivileged";
+    }
+    if (NULL != logFile) {
+        length = snprintf(NULL, 0, "--log-file=%s", logFile);
+        if (length < 0) {
+            JLI_ReportErrorMessage("Failed to calculate the length of the command option, value=%s, format=%s.", logFile, "--log-file=%s");
+            restoreStatus = -1;
+            goto freeLogLevelOption;
+        }
+        logFileOption = (char *)JLI_MemAlloc(length + 1);
+        if (NULL == logFileOption) {
+            JLI_ReportErrorMessage("Failed to allocate memory for the command option, value=%s, format=%s.", logFile, "--log-file=%s");
+            restoreStatus = -1;
+            goto freeLogLevelOption;
+        }
+        if (snprintf(logFileOption, length + 1, "--log-file=%s", logFile) < 0) {
+            JLI_ReportErrorMessage("Failed to allocate memory for the command option, value=%s, format=%s.", logFile, "--log-file=%s");
+            restoreStatus = -1;
+            goto freeLogFileOption;
+        }
+        argv[argc++] = logFileOption;
+    }
+    argv[argc] = NULL;
+    execvp(argv[0], (char * const *)argv);
+    /* If execvp returns, there was an error. */
+    restoreStatus = -1;
+freeLogFileOption:
+    if (logFileOption != NULL) {
+        JLI_MemFree((void *)logFileOption);
+        logFileOption = NULL;
+    }
+freeLogLevelOption:
+    if (logLevelOption != NULL) {
+        JLI_MemFree((void *)logLevelOption);
+        logLevelOption = NULL;
+    }
+done:
+    return restoreStatus;
+}
+
+/**
+ * Handle the restoration of the system state from a checkpoint.
+ * @param[in] argc The number of command line arguments
+ * @param[in] argv The array of command line arguments
+ */
+void
+handleCRaCRestore(int argc, char **argv)
+{
+    const char *checkpointDirectory = NULL;
+    int error = 0;
+    int parentProcessExitStatus = EXIT_SUCCESS;
+    pid_t childProcessPid = 0;
+    int logLevel = 0;
+    int childProcessExitStatus = EXIT_SUCCESS;
+    jboolean unprivilegedModeOn = JNI_FALSE;
+    const char *logFile = NULL;
+    int childProcessPidStatus = 0;
+    int childProcessPidExitStatus = 0;
+    checkpointDirectory = getCheckpointDirectory(argc, argv, &error);
+    if (-1 == error) {
+        return;
+    }
```

Comment:
A comment would be welcome here, explaining that if the restore option is not specified.